### PR TITLE
Add key/value caching for autoregressive generation

### PR DESCRIPTION
**Pull request by zygi**
Modifies model and generation code to support key/value caching. This should mostly be a no-op in terms of behavior except for some numerical stability related changes.

Tested locally. Semantic array generation only speeds up a little bit, likely because the context is small, but coarse code generation is sped up quite substantially.
Tested by generating audio with randomness held constant and confirming that very-similar-sounding results are generated.

### CODE REVIEW

**Comment by gkucsko** [comment]
amazing thanks, lots going on right now but will defo try to merge in the next couple of days. When i played with it in the past i didn't see much speedup on modern gpus with flash but maybe that's different on other gpus/cpus? or maybe i had a bug haha. did you try to benchmark on cpu? i know it's slow but might still be way faster

**Comment by wsippel** [comment]
This patch makes a huge difference for me, on a Radeon 7900XTX (running ROCm 5.5rc4). I tested this with my extension for the Oobabooga LLM frontend, and it's about 4.5 times faster with the k/v cache enabled (including text generation, which obviously doesn't change, so it's actually even faster). I went from roughly one minute for 15 seconds of audio to realtime generation.

**Comment by zygi** [comment]
> amazing thanks, lots going on right now but will defo try to merge in the next couple of days. When i played with it in the past i didn't see much speedup on modern gpus with flash but maybe that's different on other gpus/cpus? or maybe i had a bug haha. did you try to benchmark on cpu? i know it's slow but might still be way faster

I haven't tested cpu yet. Currently profiling (on gpu) tells me a significant fraction of the time is spent on python/torch overhead so I'm working on squeezing the code through `torch.compile` to see if that improves things.

**Comment by gkucsko** [comment]
much appreciated!

**Comment by zygi** [comment]
**CPU benchmarks**: there's a massive improvement. Time to generate an example sentence drops from ~30min to 2min. Tested on an i9-13900K, with the following code:
```python
from bark import SAMPLE_RATE, generate_audio
from IPython.display import Audio

text_prompt = """
     Hello, my name is Suno. And, uh — and I like pizza. [laughs] 
     But I also have other interests such as playing tic tac toe.
"""
audio_array = generate_audio(text_prompt, history_prompt="en_speaker_1", use_kv_caching=True) # or False
Audio(audio_array, rate=SAMPLE_RATE)
```

**torch.compile**: bad news, unfortunately, Torch doesn't properly support the combination of torch._dynamo + dynamic shapes + AMP. I needed to switch to nightly to even get it to show signs of life, but even then it's not functioning well. Dropping either AMP or dynamic shapes makes it work, but results in a net negative change in performance. In conclusion, no torch.compile for now unless someone's willing to do significant refactoring.

**Comment by SimFG** [comment]
maybe you can try to use [GPTCache](https://github.com/zilliztech/GPTCache). It can provide similar search, customize embedding function, provide storage function, and customize similar evaluation function for cached results, which can control cache more flexibly. 🤗 Hope GPTCache can help you.

File: bark/model.py [hunk 1]
```diff
@@ -52,22 +52,42 @@ def forward(self, x):
...
+        else:
+            present = None
+
         # causal self-attention; Self-attend: (B, nh, T, hs) x (B, nh, hs, T) -> (B, nh, T, T)
         if self.flash:
             # efficient attention using Flash Attention CUDA kernels
-            y = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=self.dropout, is_causal=True)
+            if layer_past is not None:
+                # in theory the attention is still causal but because we're computing it incrementally,
+                # the last query can attend on all previous keys/values, which which is equivalent to non-causal
```
Comment (gkucsko) [inline bark/model.py:73]: shouldn't it not matter in that case if we put causal/non-causal? honestly asking

File: bark/model.py [hunk 2]
```diff
@@ -52,22 +52,42 @@ def forward(self, x):
...
+        if layer_past is not None:
+            past_key = layer_past[0]
+            past_value = layer_past[1]
+            k = torch.cat((past_key, k), dim=-2)
+            v = torch.cat((past_value, v), dim=-2)
+
+        FULL_T = k.shape[-2]
+
+        if use_cache is True:
+            present = (k, v)
```
Comment (gkucsko) [inline bark/model.py:64]: maybe rename `present` to `current_kv` or even just `kv`?

File: bark/model.py [hunk 3]
```diff
@@ -52,22 +52,42 @@ def forward(self, x):
         q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
         v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
 
+        if layer_past is not None:
```
Comment (gkucsko) [inline bark/model.py:55]: can we homogenize the variable naming a bit for consistency? for examples `layer_past` to `past_kv`?

**Review by gkucsko — COMMENTED** [review]
amazing thanks. just some minor comments around variable renaming and please rebase with head to resolve (very minor) conflicts

File: bark/model.py (hunk 1 above)
Comment (zygi) [reply bark/model.py]: It does, updated the comment to better explain why, also empirically checked that setting it to True here breaks generation.

File: bark/model.py (hunk 3 above)
Comment (zygi) [reply bark/model.py]: Done.

**Comment by gkucsko** [comment]
boom, amazing thanks! will prob make default in another commit.

**Comment by iv2985** [comment]
> boom, amazing thanks! will prob make default in another commit.

@gkucsko

- Are the current huggingface models these accelerated ones or the original slower ones?
- Is there any way to use kv caching or assign min_eos_p from the standard transformers library rather than the bark API?

**Pull request merged**